From 26d80210a0b9632d3e11d509fc65142a3ac55d69 Mon Sep 17 00:00:00 2001
From: Olivier Brunel <jjk@jjacky.com>
Date: Sun, 28 Sep 2014 15:42:11 +0200
Subject: [PATCH 3/5] treeview: Change how toggle in handled on rubber banding

Instead of ignoring the first row, assuming it was handled on button-press
(i.e. selected state toggled with Ctrl held) , we "undo" the toggle on
rubber-band-active.

This results in the same effect for default behaviors, but allows rubber
banding operations done manually (i.e. via custom test-rubber-banding)
to behave correctly/as expected as well via custom rubber-band-active.

Signed-off-by: Olivier Brunel <jjk@jjacky.com>
---
 gtk/gtktreeview.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index 76187e9..76b1d27 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -926,6 +926,7 @@ static void gtk_tree_view_drag_gesture_end                  (GtkGestureDrag *ges
 static gboolean gtk_tree_view_test_rubber_banding           (GtkTreeView    *tree_view,
                                                              gint            bin_x,
                                                              gint            bin_y);
+static void gtk_tree_view_rubber_banding_active             (GtkTreeView    *tree_view);
 
 static guint tree_view_signals [LAST_SIGNAL] = { 0 };
 
@@ -1005,6 +1006,7 @@ gtk_tree_view_class_init (GtkTreeViewClass *class)
   class->select_cursor_parent = gtk_tree_view_real_select_cursor_parent;
   class->start_interactive_search = gtk_tree_view_start_interactive_search;
   class->test_rubber_banding = gtk_tree_view_test_rubber_banding;
+  class->rubber_banding_active = gtk_tree_view_rubber_banding_active;
 
   /* Properties */
 
@@ -3375,6 +3377,23 @@ gtk_tree_view_test_rubber_banding (GtkTreeView *tree_view,
 }
 
 static void
+gtk_tree_view_rubber_banding_active (GtkTreeView *tree_view)
+{
+  GtkTreeViewPrivate *priv = tree_view->priv;
+
+  /* if Ctrl was held, because we don't skip the first row anymore in
+   * gtk_tree_view_update_rubber_band_selection_range() we need to adjust its
+   * selected state, to keep things consistent/as expected. */
+  if (priv->rubber_band_modify && priv->rubber_band_start_node)
+    {
+      if (GTK_RBNODE_FLAG_SET (priv->rubber_band_start_node, GTK_RBNODE_IS_SELECTED))
+        GTK_RBNODE_UNSET_FLAG (priv->rubber_band_start_node, GTK_RBNODE_IS_SELECTED);
+      else
+        GTK_RBNODE_SET_FLAG (priv->rubber_band_start_node, GTK_RBNODE_IS_SELECTED);
+    }
+}
+
+static void
 gtk_tree_view_drag_gesture_begin (GtkGestureDrag *gesture,
                                   gdouble         start_x,
                                   gdouble         start_y,
@@ -4412,9 +4431,6 @@ gtk_tree_view_update_rubber_band_selection_range (GtkTreeView *tree_view,
 						 gboolean     skip_start,
 						 gboolean     skip_end)
 {
-  if (start_node == end_node)
-    return;
-
   /* We skip the first node and jump inside the loop */
   if (skip_start)
     goto skip_first;
-- 
2.4.0

